cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedTable.java

```diff
@@ -518,36 +558,49 @@ public void write(byte[] bytes, Put put) {
     put(put);
   }
 
-  private class IndexScanner implements Scanner {
+  private abstract class AbstractIndexScanner implements Scanner {
     // scanner over index table
     private final Scanner baseScanner;
-    private final byte[] rowKeyPrefix;
+    private final byte[] column;
 
-    public IndexScanner(Scanner baseScanner, byte[] rowKeyPrefix) {
+    public AbstractIndexScanner(Scanner baseScanner, byte[] column) {
       this.baseScanner = baseScanner;
-      this.rowKeyPrefix = rowKeyPrefix;
+      this.column = column;
     }
 
+    // checks if a particular column value matches a criteria defined by the implementing class
+    // return false indicates to skip the corresponding row
+    protected abstract boolean matches(byte[] columnValue);
+
     @Nullable
     @Override
     public Row next() {
       // TODO: retrieve results in batches to minimize RPC overhead (requires multi-get support in table)
-      Row dataRow = null;
       // keep going until we hit a non-null, non-empty data row, or we exhaust the index
-      while (dataRow == null) {
-        Row indexRow = baseScanner.next();
-        if (indexRow == null) {
-          // end of index
-          return null;
-        }
+      Row indexRow;
+      while ((indexRow = baseScanner.next()) != null) {
```

Comment:
I know it's convenient, but we try to avoid assignments inside conditions. Code style :)